icu4c/source/common/uresbund.cpp

```diff
@@ -2068,6 +2079,62 @@ ures_getAllItemsWithFallback(const UResourceBundle *bundle, const char *path,
     getAllItemsWithFallback(rb, value, sink, errorCode);
 }
 
+struct GetAllChildrenSink : public ResourceSink {
+
+    // Destination sink
+    ResourceSink& dest;
+    
+    // Original containing resource bundle
+    const UResourceBundle* containingBundle;
+    
+    // Table resource's key path
+    const char* tableKeyPath;
+    
+    GetAllChildrenSink(ResourceSink& dest, const UResourceBundle* containingBundle, const char* tableKeyPath)
+        : dest(dest), containingBundle(containingBundle), tableKeyPath(tableKeyPath) {}
+    virtual ~GetAllChildrenSink();
+
+    virtual void put(const char *key, ResourceValue &value, UBool isRoot,
+            UErrorCode &errorCode) {
+        ResourceTable itemsTable = value.getTable(errorCode);
+        if (U_FAILURE(errorCode)) { return; }
+        for (int32_t i = 0; itemsTable.getKeyAndValue(i, key, value); ++i) {
+            if (value.getType() == URES_ALIAS) {
+                // if the current entry in the table is an alias, re-fetch it using ures_getByKeyWithFallback():
+                // this will follow the alias (and any aliases it points to) and bring back the real value
+                CharString path;
+                path.append(tableKeyPath, errorCode);
+                path.append('/', errorCode);
+                path.append(key, errorCode);
+                
+                UResourceBundle child;
+                ures_initStackObject(&child);
+                ures_getByKeyWithFallback(containingBundle, path.data(), &child, &errorCode);
```

Comment:
>This seems inefficient: It will look for, find, and read this value’s same alias resource, and then follow the alias path; but we have already gotten this far.

I agree, but I wonder how inefficient is it in practice?  I did it this way (and did some similar things in ICU-20400) because it was a lot easier, I knew it would do the right thing, and there didn't appear to be a ready-made API that just followed the alias, but I swear I also found several precedents in the code already that do the same kind of thing, leading me to believe it wasn't _that_ inefficient, or that this wouldn't happen often enough to cause meaningful performance problems.

Another thing that swayed me this direction was that the alias might _itself_ point to an alias, which would _also_ need to be followed, and the code to do that got hairy in a hurry.  But then there are your suggestions below, which would address that...

>Also, I wonder if we should hoist this part of the logic into its own internal-public helper function...

I thought about this.  It probably _is_ a good idea, but I didn't pursue it both out of a desire to get something done and out of concern that there was a "good reason" why we didn't already have something like that.

>Finally, your changes here raise the question of whether we should ever return a ResourceValue with type=URES_ALIAS. Should we always follow aliases in all ures_getAllXyz() and ResourceTable.get/find...Value()?

That's an interesting question.  It makes intuitive sense to me, but I wonder if there are any client frameworks that actually care about alias resources and are doing their own thing?  Even if so, having two sets of APIs (or an extra parameter) seems like a reasonable approach.

I think everything you're suggesting here is reasonable, but that I need to think about it and experiment a bit before committing one way or the other.  I'm going to address all of your smaller comments first and then come back to this.  I'll either come to you to discuss this more or I'll put up another PR once I've thought these issues through.

Reply:
We can punt this discussion to another ticket or another PR.

As for how inefficient, it's an obvious duplication of work for a single resource item, so if there are a lot of aliases then it could be noticeable. More generally, if there are obvious duplications, I try to get rid of them :-) (Same for unnecessary memory allocations+initializations, such as the UResourceBundle that we are throwing away in the end.)

For my first suggestion here, I was thinking only of removing the first layer, and calling some function that takes the alias URes value and follows it, not copying the whole code that may recurse.